verilog/tools/ls/verilog-language-server_test.cc

```diff
@@ -0,0 +1,510 @@
+  void SetUp() override {
+    server = std::make_unique<VerilogLanguageServer>(
+        [this](absl::string_view response) { responsestream << response; });
+
+    absl::Status status = InitializeCommunication(Server());
+    EXPECT_TRUE(status.ok()) << "Failed to read request:  " << status;
+    Response(&initialize_response);
+  }
+
+  // Currently tested instance of VerilogLanguageServer
+  std::unique_ptr<VerilogLanguageServer> server;
+
+  // Response from textDocument/initialize request - left for checking e.g
+  // server capabilities
+  std::string initialize_response;
+
+  // Stream for passing requests to the Language Server
+  std::stringstream requeststream;
+
+  // Stream for receiving responses from the Language Server
+  std::stringstream responsestream;
+};
+
+// Verifies textDocument/initialize request handling
+TEST_F(VerilogLanguageServerTest, InitializeRequest) {
+  json response = json::parse(initialize_response);
+
+  ASSERT_EQ(response["id"], 1) << "Response message ID invalid";
+  ASSERT_EQ(response["result"]["serverInfo"]["name"],
+            "Verible Verilog language server.")
+      << "Invalid Language Server name";
+}
+
+std::string DidOpenRequest(absl::string_view name, absl::string_view content) {
+  return absl::StrCat(
+      R"({ "jsonrpc": "2.0", "method": "textDocument/didOpen", "params": { "textDocument": {"uri": ")",
+      name, R"(", "text": ")", content, R"("}}})");
+}
+
+// Checks automatic diagnostics for opened file and textDocument/diagnostic
+// request for file with invalid syntax
+TEST_F(VerilogLanguageServerTest, SyntaxError) {
+  const std::string wrong_file =
+      DidOpenRequest("file://syntaxerror.sv", R"(brokenfile\n)");
+  Request(wrong_file);
+  absl::Status status = ServerStep();
+  ASSERT_TRUE(status.ok()) << "Failed to process file with syntax error:  "
+                           << status;
+  std::string response;
+  Response(&response);
+  json response_parsed = json::parse(response);
+  ASSERT_EQ(response_parsed["method"], "textDocument/publishDiagnostics")
+      << "textDocument/publishDiagnostics not received";
+  ASSERT_EQ(response_parsed["params"]["uri"], "file://syntaxerror.sv")
+      << "Diagnostics for invalid file";
+  ASSERT_TRUE(absl::StrContains(
+      response_parsed["params"]["diagnostics"][0]["message"], "syntax error"))
+      << "No syntax error found";
+
+  // query diagnostics explicitly
+  const absl::string_view diagnostic_request = R"(
+    {
+      "jsonrpc": "2.0", "id": 2, "method": "textDocument/diagnostic",
+      "params":
+      {
+        "textDocument": {"uri": "file://syntaxerror.sv"}
+      }
+    }
+  )";
+  Request(diagnostic_request);
+  status = ServerStep();
+  ASSERT_TRUE(status.ok()) << "Failed to process file with syntax error:  "
+                           << status;
+  std::string request_response;
+  Response(&request_response);
+  response_parsed = json::parse(request_response);
```

Comment:
Here is an example where the `GetResponse()` that returns the value directly would be easier. Instead of three lines, we'd only need
```c++
  response_parsed = json::parse(GetResponse());
```